### ISSUE TITLE
Reproducible Builds

### ISSUE THREAD

**Issue by IzzySoft** [gohj99/Telewatch#14]
At IzzyOnDroid (where [Telewatch is listed](https://apt.izzysoft.de/packages/com.gohj99.telewatch) – be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme here) we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Telewatch is almost reproducible, with one very little difference:

```diff
  -rw-r--r--  0.0 unx       66 b-       59 defN 1981-01-01 01:01:02 67c26c47 assets/config.properties.example
- -rw-r--r--  0.0 unx       91 b-       85 defN 1981-01-01 01:01:02 fd8663ca assets/config.properties
  -rw----     2.0 fat     1738 b-      782 defN 1981-01-01 01:01:02 2d73be70 DebugProbesKt.bin
```

As the diff shows, there's one additional file in the APK you've built which is not part of this repo. As there are otherwise no differences in the APKs, I'm not even sure it's needed to build it. I've just established RB for Telewatch (taking that file out of your APK, encode it with `base64`, and inject it into the tree here before building), but this way would mean each release needing manual work to get the RB fixed:

```yaml
build:
  - chmod +x gradlew
  - base64 -d > app/src/main/assets/config.properties <<< YXBpX2lkPTIzMzgxNDc2CmFwaV9oYXNoPTE2YTg5MTY2MzRhMmEyOWQyYTFhNzkxYTZjNzIxYzEyCkJVSUxEX0RBVEU9MjAyNDExMDIxOQpCRVRBPWZhbHNlCg==
  - ./gradlew assembleArmRelease
```

Could that file either be checked in to the repo (directly or as base64-encoded variant we could automatically fetch for our build recipe) – or entirely excluded if not needed for the build?

We'd appreciate if you could help making your build _easily_ reproducible. We've also prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that in case you're curious.

Looking forward to your reply!

**Comment by IzzySoft** [comment]
PS: I've just discovered that since v1.1.1, no new updates arrived at IzzyOnDroid. Culprit was you changed the pattern of your tag names (from `v<versionName>-\d+` to just `v<versionName>`). This is now adjusted on our end. Please remember to let us know when you change that next so updates are not delayed again :wink: 

That said, manually triggering an update (so see if our adjustments worked), I see the scanner complaining:

    ! repo/com.gohj99.telewatch_17.apk declares sensitive permission(s): android.permission.REQUEST_INSTALL_PACKAGES

Could you please explain what that is needed for? I hope it's not a self-updater? Or installing "other apps"?

**Comment by gohj99** [comment]
> PS: I've just discovered that since v1.1.1, no new updates arrived at IzzyOnDroid. Culprit was you changed the pattern of your tag names (from `v<versionName>-\d+` to just `v<versionName>`). This is now adjusted on our end. Please remember to let us know when you change that next so updates are not delayed again 😉
> 
> That said, manually triggering an update (so see if our adjustments worked), I see the scanner complaining:
> 
> ```
> ! repo/com.gohj99.telewatch_17.apk declares sensitive permission(s): android.permission.REQUEST_INSTALL_PACKAGES
> ```
> 
> Could you please explain what that is needed for? I hope it's not a self-updater? Or installing "other apps"?

Yes, it's a self-update
but it's need userclick on the update check

**Comment by gohj99** [comment]
> At IzzyOnDroid (where [Telewatch is listed](https://apt.izzysoft.de/packages/com.gohj99.telewatch) – be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme here) we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Telewatch is almost reproducible, with one very little difference:
> 
> ```diff
>   -rw-r--r--  0.0 unx       66 b-       59 defN 1981-01-01 01:01:02 67c26c47 assets/config.properties.example
> - -rw-r--r--  0.0 unx       91 b-       85 defN 1981-01-01 01:01:02 fd8663ca assets/config.properties
>   -rw----     2.0 fat     1738 b-      782 defN 1981-01-01 01:01:02 2d73be70 DebugProbesKt.bin
> ```
> 
> As the diff shows, there's one additional file in the APK you've built which is not part of this repo. As there are otherwise no differences in the APKs, I'm not even sure it's needed to build it. I've just established RB for Telewatch (taking that file out of your APK, encode it with `base64`, and inject it into the tree here before building), but this way would mean each release needing manual work to get the RB fixed:
> 
> ```yaml
> build:
>   - chmod +x gradlew
>   - base64 -d > app/src/main/assets/config.properties <<< YXBpX2lkPTIzMzgxNDc2CmFwaV9oYXNoPTE2YTg5MTY2MzRhMmEyOWQyYTFhNzkxYTZjNzIxYzEyCkJVSUxEX0RBVEU9MjAyNDExMDIxOQpCRVRBPWZhbHNlCg==
>   - ./gradlew assembleArmRelease
> ```
> 
> Could that file either be checked in to the repo (directly or as base64-encoded variant we could automatically fetch for our build recipe) – or entirely excluded if not needed for the build?
> 
> We'd appreciate if you could help making your build _easily_ reproducible. We've also prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that in case you're curious.
> 
> Looking forward to your reply!

Thank you for your feedback and for your interest! I’d like to clarify that the config.properties file is necessary for the build process. You can check this [link](https://github.gohj99.site/gohj99/Telewatch/blob/master/.github/workflows/app-build.yml) for more information: app-build.yml. The config.properties file is automatically injected during the compilation.

Additionally, if possible, I would appreciate it if you could provide a solution to help protect the API keys while still allowing for reproducible builds. This way, we can ensure both the security and reproducibility of the builds.

Thank you for your support, and I look forward to your response!

**Comment by IzzySoft** [comment]
> I’d like to clarify that the config.properties file is necessary for the build process.

I see – OK, then we have to find a way to work with this.

> I would appreciate it if you could provide a solution to help protect the API keys while still allowing for reproducible builds.

Of course: that's why we chose to implement this via base64 (and why I didn't "quote" the config file's content here). As the `assets/config.properties` ends up in _your_ APK file anyway, it's easy to extract from that – which we of course cannot prevent for the APK you provide :wink: So our recipe just holds a base64 encoded string. Sure, this can be extracted easily as well – effort is not higher or lower than to extract it from the APK.

One idea would be to have that base64 string somewhere in your repository. The difficulty of that is just one thing: `echo "BUILD_DATE=${{ env.BUILD_DATE }}" >> app/src/main/assets/config.properties` – as that is generated at build time. Though if it ends up in the base64 you use then, and we'd use the same, that would "solve itself".

So would a `config.64` file (name is just an example) checked in to your repo be something you'd consider? That's be the easiest way:

    cat config.64 | base64 -d > assets/config.local

in the recipe (and in your workflow then, too). That's how it's done by other apps as well, let me check our recipes…

* `base64 -d > local.properties <<< ...`: 3x
* `base64 -d >> gradle.properties <<< ...`: 1x

Not counting Telewatch. To put into proportion: 315 apps are currently covered by this builder. API keys and such are only needed by very few of those, and have always been dealt with this way (with the knowledge of the developers as far as I know – at least for the recipes I've set up).

> Thank you for your support

Thank you for yours! Hoping for a solution working for both of us :wink:

**Comment by IzzySoft** [comment]
> Yes, it's a self-update
> but it's need userclick on the update check

Is the update check itself opt-in, and does it explain where updates are checked for and where the APKs come from – including consequences of bypassing the additional checks in repos like IzzyOnDroid or F-Droid? Quoting from the corresponding section of the [IzzyOnDroid App Inclusion Policy](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Inclusion-Policy) (this point is identical to F-Droid's btw):

> [...] Consent means it **needs to be opt-in** (it must not be harder to decline than to accept or presented in a way users are likely to press accept without reading) and structured in a way that **clearly explains to users that they’re choosing to bypass the checks performed in this repo if they activate it.**

(emphasis added by me here).

**Comment by IzzySoft** [comment]
PS, related question: is ACRA configured opt-in, or is it enabled by default – i.e. are reports sent automatically or only after the user has confirmed to send them?

**Comment by gohj99** [comment]
> PS, related question: is ACRA configured opt-in, or is it enabled by default – i.e. are reports sent automatically or only after the user has confirmed to send them?

To ensure user experience, I will ask the user whether to enable the ACRA service at the first startup. However, before startup, the ACRA service is enabled by default.

**Comment by gohj99** [comment]
> > Yes, it's a self-update
> > but it's need userclick on the update check
> 
> Is the update check itself opt-in, and does it explain where updates are checked for and where the APKs come from – including consequences of bypassing the additional checks in repos like IzzyOnDroid or F-Droid? Quoting from the corresponding section of the [IzzyOnDroid App Inclusion Policy](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Inclusion-Policy) (this point is identical to F-Droid's btw):
> 
> > [...] Consent means it **needs to be opt-in** (it must not be harder to decline than to accept or presented in a way users are likely to press accept without reading) and structured in a way that **clearly explains to users that they’re choosing to bypass the checks performed in this repo if they activate it.**
> 
> (emphasis added by me here).

Yes, enabling self-updates is entirely voluntary for the user. They must click on the "Check for Updates" option to initiate the process, and the updates are sourced directly from the releases on our GitHub repository.

**Comment by IzzySoft** [comment]
> To ensure user experience, I will ask the user whether to enable the ACRA service when starting the app for the first time. However, before asking, the ACRA service is enabled by default.

I read that as meaning "until now it was enabled by default, but with future versions it will be clearly opt-in and only enabled when confirmed", right? Because otherwise we'll need the `Tracking` anti-feature here.

> Yes, enabling self-updates is entirely voluntary for the user. They must click on the "Check for Updates" option to initiate the process

Good, thanks! And are the implications (bypassing of checks) made clear at that place?

**Comment by gohj99** [comment]
‎> I read that as meaning "until now it was enabled by default, but with future versions it will be clearly opt-in and only enabled when confirmed", right? Because otherwise we'll need the `Tracking` anti-feature here.
No
Please allow me report by chinese
在第一次启动时，acra服务默认是启用的，这是为了防止在某些设备上因为各种原因报错导致用户无法处理是否处理是否收集信息的提示
而如果软件正常运行，会在第一时间提示用户是否允许收集用户信息，用户可以点击是或者否。但是软件出现报错，没有成功显示是否允许收集数据的页面话，会上报到服务器
如果您对此有意见，我可以修改成默认不收集数据，直到用户点击同意后再收集

**Comment by gohj99** [comment]
‎> Good, thanks! And are the implications (bypassing of checks) made clear at that place?
Yes, you can understand it.

**Comment by gohj99** [comment]
> > I’d like to clarify that the config.properties file is necessary for the build process.
> 
> I see – OK, then we have to find a way to work with this.
> 
> > I would appreciate it if you could provide a solution to help protect the API keys while still allowing for reproducible builds.
> 
> Of course: that's why we chose to implement this via base64 (and why I didn't "quote" the config file's content here). As the `assets/config.properties` ends up in _your_ APK file anyway, it's easy to extract from that – which we of course cannot prevent for the APK you provide 😉 So our recipe just holds a base64 encoded string. Sure, this can be extracted easily as well – effort is not higher or lower than to extract it from the APK.
> 
> One idea would be to have that base64 string somewhere in your repository. The difficulty of that is just one thing: `echo "BUILD_DATE=${{ env.BUILD_DATE }}" >> app/src/main/assets/config.properties` – as that is generated at build time. Though if it ends up in the base64 you use then, and we'd use the same, that would "solve itself".
> 
> So would a `config.64` file (name is just an example) checked in to your repo be something you'd consider? That's be the easiest way:
> 
> ```
> cat config.64 | base64 -d > assets/config.local
> ```
> 
> in the recipe (and in your workflow then, too). That's how it's done by other apps as well, let me check our recipes…
> 
> * `base64 -d > local.properties <<< ...`: 3x
> * `base64 -d >> gradle.properties <<< ...`: 1x
> 
> Not counting Telewatch. To put into proportion: 315 apps are currently covered by this builder. API keys and such are only needed by very few of those, and have always been dealt with this way (with the knowledge of the developers as far as I know – at least for the recipes I've set up).
> 
> > Thank you for your support
> 
> Thank you for yours! Hoping for a solution working for both of us 😉
Ok, I see
I will talk to you late

**Comment by IzzySoft** [comment]
> Please allow me report by chinese

Sure! I cannot read Chinese, but I have a good translation tool: DeepL.com :stuck_out_tongue_winking_eye:

> If you have a problem with this, I can change the default to not collect data until the user clicks yes.

That would do the trick, and no `Tracking` anti-feature would be needed, thanks!

>> Good, thanks! And are the implications (bypassing of checks) made clear at that place?

> Yes, you can understand it.

Guess I'll have to take your word for that then?

![image](https://github.com/user-attachments/assets/86c137d9-e5c4-4308-a0aa-2d3ca8c81e60)
![image](https://github.com/user-attachments/assets/b8c6d76a-d006-433e-8cef-ee304bb63227)

??? No hint on where the check happens. Just "check update". No mention of the implications of "circumventing the additional checks in the repo". So no, that doesn't meet the criteria, sorry. Could you insert a screen with information explaining that, please (it can have a "do not ask me again" checkbox of course)? Not all people are as tech-savvy to get that idea when seeing the download page.

**Comment by gohj99** [comment]
> > Please allow me report by chinese
> 
> Sure! I cannot read Chinese, but I have a good translation tool: DeepL.com 😜
> 
> > If you have a problem with this, I can change the default to not collect data until the user clicks yes.
> 
> That would do the trick, and no `Tracking` anti-feature would be needed, thanks!
> 

Ok, I will change it to not collect data until the user clicks "yes".

> > > Good, thanks! And are the implications (bypassing of checks) made clear at that place?
> 
> > Yes, you can understand it.
> 
> Guess I'll have to take your word for that then?
> 
> ![image](https://private-user-images.githubusercontent.com/6781438/382925603-86c137d9-e5c4-4308-a0aa-2d3ca8c81e60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3OTQ0MjMsIm5iZiI6MTczMDc5NDEyMywicGF0aCI6Ii82NzgxNDM4LzM4MjkyNTYwMy04NmMxMzdkOS1lNWM0LTQzMDgtYTBhYS0yZDNjYThjODFlNjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDVUMDgwODQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTdhN2NlYjExY2RiODNhZTdjOTcxZjIwYjhjMzg2NDQ2N2I2ZGQ5Mzc0M2U2OGVmM2JjOWI2ZDQzOGFhMjdhNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Wsmwjsj9fZfMZR0-iNW72bK4PxkuOkuuFJGRjIqKw3w) ![image](https://private-user-images.githubusercontent.com/6781438/382925732-b8c6d76a-d006-433e-8cef-ee304bb63227.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3OTQ0MjMsIm5iZiI6MTczMDc5NDEyMywicGF0aCI6Ii82NzgxNDM4LzM4MjkyNTczMi1iOGM2ZDc2YS1kMDA2LTQzM2UtOGNlZi1lZTMwNGJiNjMyMjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDVUMDgwODQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2UxZTQwY2NkNjlkNDRhZDA4YTg3MjNhZGQ2ZDFkNGRjZjI2ZDFlZmJkMzFhYzAzN2ZkYmMwYmNjYTMzZjc1OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hTNgLbjZyT7ViBy0lW1eBXllomQOwBpSL_kP2fIAmeg)
> 
> ??? No hint on where the check happens. Just "check update". No mention of the implications of "circumventing the additional checks in the repo". So no, that doesn't meet the criteria, sorry. Could you insert a screen with information explaining that, please (it can have a "do not ask me again" checkbox of course)? Not all people are as tech-savvy to get that idea when seeing the download page.

Ok, I will add a screen in the next update to make it clear to users that this package comes from GitHub.

**Comment by IzzySoft** [comment]
Thanks a lot!

**Comment by gohj99** [comment]
I make some changes in the version 1.2.7.
>>
Ok, I will change it to not collect data until the user clicks "yes".
>
I am solved the issues
>>
Ok, I will add a screen in the next update to make it clear to users that this package comes from GitHub.
>
I am solved the issues

**Comment by IzzySoft** [comment]
Thanks! The ACRA intro looks fine. As for update check: better, but it does not mention the implications (that checks from Iod/F-Droid/… are bypassed). Maybe append "If you installed the app from e.g. IzzyOnDroid/F-Droid, their additional checks will **not** be run."?

**Comment by gohj99** [comment]
>>
>Maybe append "If you installed the app from e.g. IzzyOnDroid/F-Droid, their additional checks will not be run."?
>
I will append on next version.

**Comment by gohj99** [comment]
> > I’d like to clarify that the config.properties file is necessary for the build process.
> 
> I see – OK, then we have to find a way to work with this.
> 
> > I would appreciate it if you could provide a solution to help protect the API keys while still allowing for reproducible builds.
> 
> Of course: that's why we chose to implement this via base64 (and why I didn't "quote" the config file's content here). As the `assets/config.properties` ends up in _your_ APK file anyway, it's easy to extract from that – which we of course cannot prevent for the APK you provide 😉 So our recipe just holds a base64 encoded string. Sure, this can be extracted easily as well – effort is not higher or lower than to extract it from the APK.
> 
> One idea would be to have that base64 string somewhere in your repository. The difficulty of that is just one thing: `echo "BUILD_DATE=${{ env.BUILD_DATE }}" >> app/src/main/assets/config.properties` – as that is generated at build time. Though if it ends up in the base64 you use then, and we'd use the same, that would "solve itself".
> 
> So would a `config.64` file (name is just an example) checked in to your repo be something you'd consider? That's be the easiest way:
> 
> ```
> cat config.64 | base64 -d > assets/config.local
> ```
> 
> in the recipe (and in your workflow then, too). That's how it's done by other apps as well, let me check our recipes…
> 
> * `base64 -d > local.properties <<< ...`: 3x
> * `base64 -d >> gradle.properties <<< ...`: 1x
> 
> Not counting Telewatch. To put into proportion: 315 apps are currently covered by this builder. API keys and such are only needed by very few of those, and have always been dealt with this way (with the knowledge of the developers as far as I know – at least for the recipes I've set up).
> 
> > Thank you for your support
> 
> Thank you for yours! Hoping for a solution working for both of us 😉

I try to revise the build file
[app-build.yml](https://github.com/gohj99/Telewatch/blob/master/.github/workflows/app-build.yml)
able to pb?

**Comment by gohj99** [comment]
And can you edit repository description to "‎Telegram for Android watch & wearos"
and edit Functions
Multi-account support
Interface font adjustment
Unread/read indicator
Connection status indicator
Viewing or downloading text, images, and videos
Saving images or videos
Text messaging
Message deletion
Support for rotating the digital crown

thank

**Comment by IzzySoft** [comment]
> able to pb?

If you mean RB: 1.2.8 shows the green shied, so it was RB, yes.

> And can you edit repository description

I can send you a PR with the metadata from here in Fastlane structures, then you can adjust descriptions and graphics to your liking. Would that be OK for you? Fastlane would be pulled along with each new release. And you're welcome to use the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) as reference. (Fastlane is the de-facto standard for such things in the Android world)

**Comment by gohj99** [comment]
> > able to pb?
> 
> If you mean RB: 1.2.8 shows the green shied, so it was RB, yes.
> 
> > And can you edit repository description
> 
> I can send you a PR with the metadata from here in Fastlane structures, then you can adjust descriptions and graphics to your liking. Would that be OK for you? Fastlane would be pulled along with each new release. And you're welcome to use the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) as reference. (Fastlane is the de-facto standard for such things in the Android world)

OK

**Comment by IzzySoft** [comment]
Could you please insert an empty line before the bullet-point list in `full_description.txt`?

```markdown
Functions :

- Multi-account support
- Interface font adjustment
- Unread/read indicator
```

That would make it render properly as markdown, giving a much nicer output. Also, adding an icon and some screenshots would be nice :wink:

>> Maybe append "If you installed the app from e.g. IzzyOnDroid/F-Droid, their additional checks will not be run."?

> I will append on next version.

Looks like you forgot, text is still unchanged.

**Comment by gohj99** [comment]
>
>Looks like you forgot, text is still unchanged.
>>
You can see commit
I change it

>
>Could you please insert an empty line before the bullet-point list in full_description.txt?
>>
Ok

**Comment by IzzySoft** [comment]
> You can see Lately commit
> I change it

I've only checked the latest APK (installed it and pushed the button), didn't see it there. The later commits don't show anything related. So which commit to look at? Ah, nevermind, found it in the develop branch – same commit as fastlane `full_description.txt` (looks fine!), thanks! Please let me know when that got merged to the master branch so I can update metadata here.

Thanks a lot!

Btw: do you also plan to add graphics (icon, screenshots) to your fastlane tree?

**Comment by gohj99** [comment]
>
>Btw: do you also plan to add graphics (icon, screenshots) to your fastlane tree?
>>


I comment in [this](https://github.com/gohj99/Telewatch/tree/develop/fastlane/metadata/android/en-US/images)

**Comment by IzzySoft** [comment]
Thanks, added to the metadata here. The Fastlane structure will be checked and pulled along with future releases then.

Looks like all done now, and we can close this issue – or did we miss anything?

**Comment by gohj99** [comment]
This question can be closed, but before closing it, I want to ask you a question: How to promote software better? Is there a website that can generate apk promotion pages?
thank you

**Comment by IzzySoft** [comment]
Sorry, I don't know about such promotion options.

**Comment by gohj99** [comment]
Hello, I recently found that the version of Telewatch on Izzysoft has not been updated. Is there any problem?

**Comment by IzzySoft** [comment]
Look like you forgot to increase `versionCode` when making the new release:

```
$ iod repo get com.gohj99.telewatch
No metadata file (yml) for 'com.susheelkaram.myread', skipping
com.gohj99.telewatch: looking for 'https://api.github.com/repos/gohj99/Telewatch/releases'
com.gohj99.telewatch: checking tag 'v1.3.0'
com.gohj99.telewatch: lastRelNo set to '1.3.0', checking for files
com.gohj99.telewatch: Upstream file date (2024-11-30 18:39) is newer than ours (2024-11-16 19:49).
com.gohj99.telewatch: returning ['1.3.0','https://github.com/gohj99/Telewatch/releases/download/v1.3.0/Telewatch-v1.3.0-arm.apk',1732988382]
com.gohj99.telewatch: 1.2.9/1.3.0, https://github.com/gohj99/Telewatch/releases: https://github.com/gohj99/Telewatch/releases/download/v1.3.0/Telewatch-v1.3.0-arm.apk
- Grabbing update for com.gohj99.telewatch: OK
repo/com.gohj99.telewatch_20.apk already exists; ignoring update for com.gohj99.telewatch, versionCode 20 already present.
```

Android uses `versionCode` to tell versions apart, so `<packageName>_<versionCode>` is a unique key. So whoever had installed v1.2.9 would not receive v1.3.0 as update as Android would simply say the same as the logs above: "this version is already present".

**Comment by gohj99** [comment]
Okay, my question

**Comment by IzzySoft** [comment]
Next release will be picked then if `versionCode` was increased, yes :wink: But better not modify the current release (rule of thumb: never replace what has alreaedy been distributed, only causes confusion – and other dragons :dragon: there).

**Comment by gohj99** [comment]
Ok
but.....
late...
I didn't see it before the replacement
sorry

**Comment by IzzySoft** [comment]
Won't be picked up then by the RB builders (which already have that tag marked as processed) – and replacing the APK that was confirmed by the builder here invalidated the confirmation, as the hashes no longer match :man_shrugging: Have to wait for the next release then, sorry – and the current release will not have the green shield up either…

**Comment by gohj99** [comment]
Thank you, I release a new version now

**Comment by IzzySoft** [comment]
Thanks! Was fetched fine by both updaters (repo & builder), and confirmed RB :smiley: Goes live with the next sync around 7 pm UTC then, as usual.
